pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -686,9 +772,45 @@ func (p *PriorityQueue) Pop() (*framework.QueuedPodInfo, error) {
 	pInfo := obj.(*framework.QueuedPodInfo)
 	pInfo.Attempts++
 	p.schedulingCycle++
+	// In flight, no move request yet.
+	p.inFlightPods[pInfo.Pod.UID] = inFlightPod{
+		startedCycle: p.schedulingCycle,
+	}
+
+	for plugin := range pInfo.UnschedulablePlugins {
+		metrics.UnschedulableReason(plugin, pInfo.Pod.Spec.SchedulerName).Dec()
+	}
+
 	return pInfo, nil
 }
 
+// Done must be called for pod returned by Pop. This allows the queue to
+// keep track of which pods are currently being processed.
+func (p *PriorityQueue) Done(pod types.UID) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	p.done(pod)
+}
+
+func (p *PriorityQueue) done(pod types.UID) {
+	if _, ok := p.inFlightPods[pod]; !ok {
+		// This Pod is already done()ed.
+		return
+	}
+	delete(p.inFlightPods, pod)
+
+	// remove events which is only referred from this Pod
+	// so that the receivedEvents map doesn't grow infinitely.
+	for event := p.receivedEvents.Front(); event != nil; event = event.Next() {
+		e := event.Value.(clusterEvent)
+		e.inFlightPodsNum--
+		if e.inFlightPodsNum == 0 {
```

Comment:
nit: for defensive programming:

```suggestion
		if e.inFlightPodsNum <= 0 {
```